src/weights.jl

```diff
@@ -1,19 +1,35 @@
 
 ###### Weight vector #####
 
-immutable WeightVec{W,Vec<:RealVector}
-    values::Vec
-    sum::W
-end
+if VERSION < v"0.6.0-dev.2123"
+    immutable WeightVec{S<:Real, T<:Real, V<:RealVector} <: RealVector{T}
+        values::V
+        sum::S
+    end
 
-"""
-    WeightVec(vs, [wsum])
+    function WeightVec{S<:Real, V<:RealVector}(vs::V, s::S)
+        return WeightVec{S, eltype(vs), V}(vs, s)
+    end
 
-Construct a `WeightVec` with weight values `vs` and sum of weights `wsum`.
-If omitted, `wsum` is computed.
-"""
-WeightVec{Vec<:RealVector,W<:Real}(vs::Vec,wsum::W) = WeightVec{W,Vec}(vs, wsum)
-WeightVec(vs::RealVector) = WeightVec(vs, sum(vs))
+    function WeightVec{V<:RealVector}(vs::V)
+        sum_ = sum(vs)
+        return WeightVec{typeof(sum_), eltype(vs), V}(vs, sum_)
+    end
+else
+    immutable WeightVec{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractVector{T}
+        values::V
+        sum::S
+    end
+
+    function WeightVec{S<:Real, T<:Real, V<:AbstractVector{T}}(vs::V, s::S)
+        return WeightVec{S, T, V}(vs, s)
+    end
+
+    function WeightVec{T<:Real, V<:AbstractVector{T}}(vs::V)
+        sum_ = sum(vs)
```

Comment:
This is mostly just habit as I try to avoid using single character variable names. For math and stats code I tend to use the name of the function name with an underscore as my variable names if I can't think of anything better.